File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/CoinJar.java
package com.mobnetic.coinguardian.model.market;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;

public class CoinJar extends Market {

	private final static String NAME = "CoinJar";
	private final static String TTS_NAME = "Coin Jar";
	private final static String URL = "https://api.coinjar.com/v3/exchange_rates";
	private final static String URL_CURRENCY_PAIRS = URL;

	public CoinJar() {
		super(NAME, TTS_NAME, null);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return URL;
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONObject ratesJsonObject = jsonObject.getJSONObject("exchange_rates");
		final JSONObject pairJsonObject = ratesJsonObject.getJSONObject(checkerInfo.getCurrencyPairId());
		ticker.bid = pairJsonObject.getDouble("bid");
		ticker.ask = pairJsonObject.getDouble("ask");
		ticker.last = pairJsonObject.getDouble("midpoint");
	}
	
	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL_CURRENCY_PAIRS;
	}

	@Override
	protected void parseCurrencyPairsFromJsonObject(int requestId, JSONObject jsonObject, List<CurrencyPairInfo> pairs) throws Exception {
		final JSONObject ratesJsonObject = jsonObject.getJSONObject("exchange_rates");
		final JSONArray namesJsonArray = ratesJsonObject.names();

		for(int i=0; i< namesJsonArray.length(); ++i) {
			final String symbol = namesJsonArray.getString(i);
			final JSONObject pairJsonObject = ratesJsonObject.getJSONObject(symbol);
			pairs.add(new CurrencyPairInfo(
					pairJsonObject.getString("base_currency"),
					pairJsonObject.getString("counter_currency"),
					symbol
			));
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/CoinJar.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.util.forEachName
import org.json.JSONObject

class CoinJar : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return URL
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        jsonObject
            .getJSONObject("exchange_rates")
            .getJSONObject(checkerInfo.currencyPairId!!).also {
                ticker.bid = it.optDouble("bid", ticker.bid)
                ticker.ask = it.optDouble("ask", ticker.ask)
                ticker.last = it.getDouble("midpoint")
            }
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        jsonObject
            .getJSONObject("exchange_rates")
            .forEachName{ symbol, item ->
                pairs.add(CurrencyPairInfo(
                    item.getString("base_currency"),
                    item.getString("counter_currency"),
                    symbol
                ))
            }
    }

    companion object {
        private const val NAME = "CoinJar Prices"
        private const val TTS_NAME = "Coin Jar prices"
        private const val URL = "https://api.coinjar.com/v3/exchange_rates"
        private const val URL_CURRENCY_PAIRS = URL
    }
}